File: Engine/src/main/java/org/etieskrill/engine/util/DisposableLoader.java
package org.etieskrill.engine.util;

import org.etieskrill.engine.Disposable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public abstract class DisposableLoader<T extends Disposable> extends Loader<T> implements Disposable {

    private static final Logger logger = LoggerFactory.getLogger(DisposableLoader.class);

    @Override
    public void dispose() {
        logger.debug("Disposing loader'{}'", getLoaderName());
        map.values().forEach(T::dispose);
    }

}

File: Engine/src/main/java/org/etieskrill/engine/util/DisposableLoader.kt
package org.etieskrill.engine.util

import io.github.oshai.kotlinlogging.KotlinLogging
import org.etieskrill.engine.common.Disposable

private val logger = KotlinLogging.logger {}

abstract class DisposableLoader<T : Disposable> : Loader<T>(), Disposable {

    override fun dispose() {
        logger.debug { "Disposing loader $loaderName" }
        map.values.forEach { it.dispose() }
    }

}
